### PR TITLE
rsemanager: _get_possible_protocols Fix #137

### LINKED ISSUE #137: rsemanager: _get_possible_protocols call raises exception while it should not
Motivation
----------
In the _get_possible_protocols call the filtering (tbr) of protocols removes too many protocols, as the list is essentially done twice (per domain) and it looks like more is removed than is there, thus an exception is raised.

Modification
------------
Make proper filtering.
Split the candidates in two lists per domain and do the filtering smarter.

Expected behavior
-----------------
Exception should not be raised.

### PR DESCRIPTION
**Pull request by vingar**

### CODE REVIEW

File: lib/rucio/rse/rsemanager.py [hunk 1]
```diff
@@ -68,31 +68,34 @@ def get_rse_info(rse, session=None):
...
 
-    for d in rse_settings['domain']:
-        if d not in utils.rse_supported_protocol_domains():
-            raise exception.RSEProtocolDomainNotSupported('Domain %s is not supported by Rucio.' % d)
-
     # convert scheme to list, if given as string
     if scheme and not isinstance(scheme, list):
         scheme = scheme.split(',')
 
     tbr = []
```
Comment (TWAtGH) [inline lib/rucio/rse/rsemanager.py]: Maybe declaring tbr as set could do the trick? Or just converting tbr to a set before using len below

File: lib/rucio/rse/rsemanager.py [hunk 2]
```diff
@@ -68,31 +68,34 @@ def get_rse_info(rse, session=None):
...
+        if d not in utils.rse_supported_protocol_domains():
+            raise exception.RSEProtocolDomainNotSupported('Domain %s is not supported'
+                                                          ' by Rucio.' % d)
+        for protocol in candidates:
+            # Check if scheme given and filter if so
+            if scheme and protocol['scheme'] not in scheme:
+                if protocol not in tbr:
+                    tbr.append(protocol)
+                continue
+            for d in rse_settings['domain']:
```
Comment (cserf) [inline lib/rucio/rse/rsemanager.py]: ??? You already looped over the domain before

File: lib/rucio/rse/rsemanager.py (hunk 1 above)
Comment (vingar) [reply lib/rucio/rse/rsemanager.py]: TypeError: unhashable type: 'dict'

**Comment by vingar** [comment]
tested with that for now:
`from rucio.rse import rsemanager as rsemgr

rse_settings =  {'rse_type': 'DISK', 'domain': ['lan', 'wan'], 'availability_delete': True, 'delete_protocol': 1, 'rse': u'RAL-LCG2_DATADISK', 'deterministic': True, 'write_protocol': 1, 'read_protocol': 1, 'staging_area': False, 'availability_write': True, 'volatile': False, 'availability_read': True, 'credentials': None, 'id': '20cf88db5d684958866a484c2221fccb', 'protocols': [{'extended_attributes': None, 'hostname': u'atlas-xrd.gridpp.rl.ac.uk', 'prefix': u'//castor/ads.rl.ac.uk/prod/atlas/stripInput/atlasdatadisk/rucio/', 'domains': {'wan': {'read': 2, 'write': 0, 'third_party_copy': None, 'delete': 2}, 'lan': {'read': 0, 'write': 0, 'delete': 2}}, 'scheme': u'root', 'port': 1094, 'impl': u'rucio.rse.protocols.gfal.Default'}, {'extended_attributes': None, 'hostname': u'catlasdlf.ads.rl.ac.uk', 'prefix': u'//castor/ads.rl.ac.uk/prod/atlas/stripInput/atlasdatadisk/rucio/', 'domains': {'wan': {'read': 0, 'write': 0, 'third_party_copy': None, 'delete': 0}, 'lan': {'read': 1, 'write': 0, 'delete': 0}}, 'scheme': u'root', 'port': 1094, 'impl': u'rucio.rse.protocols.gfal.Default'}, {'extended_attributes': {u'space_token': u'ATLASDATADISK', u'web_service_path': u'/srm/managerv2?SFN='}, 'hostname': u'srm-atlas.gridpp.rl.ac.uk', 'prefix': u'/castor/ads.rl.ac.uk/prod/atlas/stripInput/atlasdatadisk/rucio/', 'domains': {'wan': {'read': 1, 'write': 1, 'third_party_copy': 1, 'delete': 1}, 'lan': {'read': 0, 'write': 1, 'delete': 1}}, 'scheme': u'srm', 'port': 8443, 'impl': u'rucio.rse.protocols.gfal.Default'}]}

print rsemgr._get_possible_protocols(rse_settings, 'read')`

**Comment by vingar** [comment]
the protocol is finally excluded if disabled (0) on all domains

File: lib/rucio/rse/rsemanager.py (hunk 1 above)
Comment (TWAtGH) [reply lib/rucio/rse/rsemanager.py]: Ok sorry. I was told protocol would be a string

Comment (vingar) [reply lib/rucio/rse/rsemanager.py]: scheme yes but no worry the code there is ...... interesting !

**Comment by bari12** [comment]
Can you add the test permanently?

**Review by bari12 — APPROVED** [review]

**Pull request merged**